hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/SparkBaseIndexSupport.scala

```diff
@@ -173,7 +174,13 @@ abstract class SparkBaseIndexSupport(spark: SparkSession,
       var recordKeyQueries: List[Expression] = List.empty
       var compositeRecordKeys: List[String] = List.empty
       val recordKeyOpt = getRecordKeyConfig
-      val isComplexRecordKey = recordKeyOpt.map(recordKeys => recordKeys.length).getOrElse(0) > 1
+      val isComplexRecordKey = {
+        val fieldCount = recordKeyOpt.map(recordKeys => recordKeys.length).getOrElse(0)
+        // Consider as complex if:
+        // 1. Multiple fields (> 1), OR
+        // 2. Using complex key generator with single record key field
+        fieldCount > 1 || isComplexKeyGeneratorWithSingleRecordKeyField(metaClient.getTableConfig)
```

Comment:
the write config flag is missing?

Reply:
This is on the reader path, so I'm thinking that maybe for tables generated by releases with breaking change, we avoid using data skipping if the key encoding cannot be derived?

Reply:
> we avoid using data skipping if the key encoding cannot be derived

It will yield wrong query results here, a match case would become mismatch. or we doc it well in some places.